### PR TITLE
clientv3: add pinned() method to 'balancer'

### PR DESCRIPTION
**Pull request by gyuho**

### CODE REVIEW

**Comment by gyuho** [comment]
To debug https://github.com/coreos/etcd/issues/8660.

File: clientv3/balancer.go [hunk 1]
```diff
@@ -42,6 +42,7 @@ type balancer interface {
 
 	endpoint(host string) string
 	endpoints() []string
+	pinned() string
```
Comment (xiang90) [inline clientv3/balancer.go]: doc string?

File: clientv3/retry.go [hunk 2]
```diff
@@ -58,12 +58,13 @@ func (c *Client) newRetryWrapper(isStop retryStopErrFunc) retryRpcFunc {
...
 				return c.ctx.Err()
 			}
+			pinned := c.balancer.pinned()
 			err := f(rpcCtx)
 			if err == nil {
 				return nil
 			}
 			if logger.V(4) {
-				logger.Infof("clientv3/retry: retry for error %v", err)
+				logger.Infof("clientv3/retry: retry for error %v (%s)", err, pinned)
```
Comment (xiang90) [inline clientv3/retry.go]: retry on pined endpoint (%s) due to error %v

File: clientv3/retry.go [hunk 3]
```diff
@@ -86,12 +87,13 @@ func (c *Client) newRetryWrapper(isStop retryStopErrFunc) retryRpcFunc {
...
 	return func(rpcCtx context.Context, f rpcFunc) error {
 		for {
+			pinned := c.balancer.pinned()
 			err := f(rpcCtx)
 			if err == nil {
 				return nil
 			}
 			if logger.V(4) {
-				logger.Infof("clientv3/auth-retry: retry for error %v", err)
+				logger.Infof("clientv3/auth-retry: retry for error %v (%s)", err, pinned)
```
Comment (xiang90) [inline clientv3/retry.go]: retry on pined endpoint (%s) due to error %v

**Comment by xiang90** [comment]
this is just for debugging, right?

File: clientv3/retry.go (hunk 3 above)
Comment (gyuho) [reply clientv3/retry.go]: @xiang90 That endpoint is errored, so it's unpinned by this line. So it's possible that we retry on different endpoint?

**Comment by gyuho** [comment]
@xiang90 Yes, this is for balancer debugging. But in the following PR, I will add gray-listing the timed out endpoint.

File: clientv3/retry.go (hunk 3 above)
Comment (xiang90) [reply clientv3/retry.go]: or let's say ` error %s on pined endpoint (%s))

basically only (%s) is not descriptive.

**Comment by gyuho** [comment]
@xiang90 PTAL. Fixed. Thanks.

**Comment by xiang90** [comment]
lgtm

**Pull request merged**